sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -2652,5 +2631,169 @@ def my_pipeline(x: Input[Artifact] = Artifact(
                 comp()
 
 
+class TestIllegalFanInCollection(unittest.TestCase):
+
+    def test_missing_collected_with_correct_annotation(self):
+        from typing import List
+
+        from kfp import dsl
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        # this error will only be surfaced if the annotation is correct,
+        # since the type-checker would catch the type LIST v primitive type
+        # mismatch first
+        with self.assertRaisesRegex(
+                type_utils.InconsistentTypeException,
+                'Argument type "NUMBER_INTEGER" is incompatible with the input type "LIST"'
+        ):
+
+            @dsl.pipeline
+            def math_pipeline() -> int:
+                with dsl.ParallelFor([1, 2, 3]) as f:
+                    t = double(num=f)
+
+                return add(nums=t.output).output
+
+    def test_missing_collected_with_incorrect_annotation(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: int) -> int:
+            return nums
+
+        with self.assertRaisesRegex(
+                RuntimeError,
+                r'Tasks cannot depend on an upstream task inside a ParallelFor that is not a common ancestor of both tasks\. Task add depends on upstream task double\. Did you mean to use dsl\.Collected to fan-in outputs from a ParallelFor sub-DAG?'
+        ):
+
+            @dsl.pipeline
+            def math_pipeline() -> int:
+                with dsl.ParallelFor([1, 2, 3]) as f:
+                    t = double(num=f)
+
+                return add(nums=t.output).output
+
+    def test_producer_condition_legal1(self):
+        from kfp import dsl
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        @dsl.pipeline
+        def math_pipeline(text: str) -> int:
+            with dsl.Condition(text == 'text'):
+                with dsl.ParallelFor([1, 2, 3]) as f:
+                    t = double(num=f)
+
+                return add(nums=dsl.Collected(t.output)).output
+
+    def test_producer_condition_legal2(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        @dsl.pipeline
+        def my_pipeline(text: str = ''):
+            with dsl.ParallelFor([1, 2, 3]) as f:
+                with dsl.Condition(text == 'text'):
+                    t = double(num=f)
+
+            with dsl.Condition(text == 'text'):
```

Comment:
What if the condition was associated with loop parameter? How would user collect, or if they could collect them after the loop?
In real world scenarios, the condition should probably depends on loop parameters, otherwise, to implement the logic you have, it would be more nature and efficient to write the condition outside the loop:
```python
with dsl.Condition(text == 'text'):
    with dsl.ParallelFor([1, 2, 3]) as f:
        t = double(num=f)
```

Reply:
I believe [this test case](https://github.com/kubeflow/pipelines/blob/f51729ef1665eeafe35aa1fbd92529a9dafca833/sdk/python/test_data/pipelines/parallelfor_fan_in/parameters_with_conditional_consumer.py#L20) covers the pipeline logic you're desribing.